src/compiler/parser.ts

```diff
@@ -2739,49 +2739,50 @@ namespace ts {
-            }
-        }
-
-        function parseArrayTypeOrHigher(): TypeNode {
-            let type = parseJSDocPostfixTypeOrHigher();
-            while (!scanner.hasPrecedingLineBreak() && parseOptional(SyntaxKind.OpenBracketToken)) {
-                if (isStartOfType()) {
-                    const node = <IndexedAccessTypeNode>createNode(SyntaxKind.IndexedAccessType, type.pos);
-                    node.objectType = type;
-                    node.indexType = parseType();
-                    parseExpected(SyntaxKind.CloseBracketToken);
-                    type = finishNode(node);
+        function parsePostfixTypeOrHigher(): TypeNode {
+            let kind: SyntaxKind | undefined;
+            let type = parseNonArrayType();
+            while (!scanner.hasPrecedingLineBreak() && (kind = getPostfixTypeKind(token()))) {
+                if (kind === SyntaxKind.ArrayType) {
+                    parseExpected(SyntaxKind.OpenBracketToken);
+                    if (isStartOfType()) {
+                        const node = createNode(SyntaxKind.IndexedAccessType, type.pos) as IndexedAccessTypeNode;
+                        node.objectType = type;
+                        node.indexType = parseType();
+                        parseExpected(SyntaxKind.CloseBracketToken);
+                        type = finishNode(node);
+                    }
+                    else {
+                        const node = createNode(SyntaxKind.ArrayType, type.pos) as ArrayTypeNode;
+                        node.elementType = type;
+                        parseExpected(SyntaxKind.CloseBracketToken);
+                        type = finishNode(node);
+                    }
                 }
                 else {
-                    const node = <ArrayTypeNode>createNode(SyntaxKind.ArrayType, type.pos);
-                    node.elementType = type;
-                    parseExpected(SyntaxKind.CloseBracketToken);
-                    type = finishNode(node);
+                    nextToken();
+                    const postfix = createNode(kind, type.pos) as JSDocOptionalType | JSDocNonNullableType | JSDocNullableType;
+                    postfix.type = type;
+                    type = finishNode(postfix);
                 }
             }
             return type;
         }
 
+        function getPostfixTypeKind(tokenKind: SyntaxKind): SyntaxKind | undefined {
+            switch (tokenKind) {
+                case SyntaxKind.EqualsToken:
+                    // only parse postfix = inside jsdoc, because it's ambiguous elsewhere
+                    return contextFlags & NodeFlags.JSDoc ? SyntaxKind.JSDocOptionalType : undefined;
+                case SyntaxKind.ExclamationToken:
+                    return SyntaxKind.JSDocNonNullableType;
+                case SyntaxKind.QuestionToken:
+                    return SyntaxKind.JSDocNullableType;
+                case SyntaxKind.OpenBracketToken:
+                    return SyntaxKind.ArrayType;

```

Comment:
This is slightly misleading - in actuality, you are potentially parsing out an `IndexedAccessTypeNode`. Is it fair to say that this function determines gives you a `JSDocPostfixType`, a `TypeScriptPostfixType`, or `None`?

Reply:
What if you just inlined this `switch`/`case` to the function itself?

Reply:
Inlining makes the jsdoc cases really inconvenient because the mapping between tokens and jsdoc types has to happen in the switch, but the `createNode` code is shared between all three. I could separate out just the token &rarr; jsdoc-type mapping code, but that would essentially **duplicate** the switch.